File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/bills/coupon_controller.py
import logging
import inspect
from django.contrib import messages
from django.contrib.auth.decorators import login_required
from django.core import signing
from django.core.exceptions import PermissionDenied
from django.http import HttpRequest, HttpResponse, JsonResponse
from django.shortcuts import get_object_or_404, redirect, render
from django.urls import reverse
from django.utils.decorators import method_decorator
from ....Models.bills.coupon import Coupon
from ....Models.bills.user_coupon import UserCoupon
from ....Models.planning.plan import Plan
from ....Models.utils.utility import Utility
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_permission_denial, default_undefined_exception
from .._traits.controller import Controller

logger = logging.getLogger(__name__)

class CouponController(Controller):

    @method_decorator(login_required)
    @classmethod
    def index(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('manage coupon'):
                raise PermissionDenied('User lacks permission: manage coupon')
            coupons = Coupon.objects.all()
            return render(request, 'coupon/index.html', {'coupons': coupons})
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @method_decorator(login_required)
    @classmethod
    def create(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('create coupon'):
                raise PermissionDenied('User lacks permission: create coupon')
            return render(request, 'coupon/create.html', {})
        except PermissionDenied as err:
            return default_permission_denial(
                request,
                err=err,
                ref=REF,
                logger=logger,
                json={'error': 'Permission denied.'}
            )
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @method_decorator(login_required)
    @classmethod
    def store(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('create coupon'):
                raise PermissionDenied('User lacks permission: create coupon')
            name = request.POST.get('name')
            discount = request.POST.get('discount')
            limit = request.POST.get('limit')
            if not name or not discount or not limit:
                messages.error(request, "Name, discount and limit are required.")
                return redirect(get_redirect_url(request))
            try:
                discount = float(discount)
                limit = int(limit)
            except Exception:
                messages.error(request, "Discount and limit must be numeric.")
                return redirect(get_redirect_url(request))
            if not request.POST.get('manualCode') and not request.POST.get('autoCode'):
                messages.error(request, "Coupon code is required.")
                return redirect(get_redirect_url(request))
            coupon = Coupon(
                name=name,
                discount=discount,
                limit=limit,
                code=(request.POST.get('manualCode').upper()
                      if request.POST.get('manualCode')
                      else request.POST.get('autoCode'))
            )
            coupon.save()
            messages.success(request, "Coupon successfully created.")
            return redirect(reverse('coupons_index'))
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @method_decorator(login_required)
    @classmethod
    def show(cls, request: HttpRequest, coupon_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            coupon = get_object_or_404(Coupon, pk=coupon_id)
            user_coupons = UserCoupon.objects.filter(coupon=coupon.id).select_related('user_detail')
            return render(request, 'coupon/view.html', {'userCoupons': user_coupons})
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @method_decorator(login_required)
    @classmethod
    def edit(cls, request: HttpRequest, coupon_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('edit coupon'):
                raise PermissionDenied('User lacks permission: edit coupon')
            coupon = get_object_or_404(Coupon, pk=coupon_id)
            return render(request, 'coupon/edit.html', {'coupon': coupon})
        except PermissionDenied as err:
            return default_permission_denial(
                request,
                err=err,
                ref=REF,
                logger=logger,
                json={'error': 'Permission denied.'}
            )
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @method_decorator(login_required)
    @classmethod
    def update(cls, request: HttpRequest, coupon_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('edit coupon'):
                raise PermissionDenied('User lacks permission: edit coupon')
            name = request.POST.get('name')
            discount = request.POST.get('discount')
            limit = request.POST.get('limit')
            code = request.POST.get('code')
            if not name or not discount or not limit or not code:
                messages.error(request, "Name, discount, limit and code are required.")
                return redirect(get_redirect_url(request))
            try:
                discount = float(discount)
                limit = int(limit)
            except Exception:
                messages.error(request, "Discount and limit must be numeric.")
                return redirect(get_redirect_url(request))
            coupon = get_object_or_404(Coupon, pk=coupon_id)
            coupon.name = name
            coupon.discount = discount
            coupon.limit = limit
            coupon.code = code
            coupon.save()
            messages.success(request, "Coupon successfully updated.")
            return redirect(reverse('coupons_index'))
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @method_decorator(login_required)
    @classmethod
    def destroy(cls, request: HttpRequest, coupon_id: int) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            if not request.user.has_perm('delete coupon'):
                raise PermissionDenied('User lacks permission: delete coupon')
            coupon = get_object_or_404(Coupon, pk=coupon_id)
            coupon.delete()
            messages.success(request, "Coupon successfully deleted.")
            return redirect(reverse('coupons_index'))
        except PermissionDenied as err:
            return default_permission_denial(request, err=err, ref=REF, logger=logger)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @method_decorator(login_required)
    @classmethod
    def apply_coupon(cls, request: HttpRequest) -> JsonResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            plan = Plan.objects.filter(pk=signing.loads(request.POST.get('plan_id'))).first()
            if not request.user.has_perm('apply coupon'):
                raise PermissionDenied('User lacks permission: apply coupon')
            if plan and request.POST.get('coupon', '') != '':
                original_price = cls.format_price(plan.price)
                coupons = Coupon.objects.filter(code=request.POST.get('coupon').upper(), is_active=1).first()
                if coupons and coupons.limit == coupons.used_coupon():
                    return JsonResponse({
                        'is_success': False,
                        'final_price': original_price,
                        'price': format(plan.price, f".{Utility.get_val_by_name('decimal_number')}f"),
                        'message': "This coupon code has expired."
                    })
                if coupons:
                    discount_amount = (plan.price / 100) * coupons.discount
                    return JsonResponse({
                        'is_success': True,
                        'discount_price': '-' + cls.format_price(discount_amount),
                        'final_price': cls.format_price(plan.price - discount_amount),
                        'price': format(plan.price - discount_amount, f".{Utility.get_val_by_name('decimal_number')}f"),
                        'message': "Coupon code has applied successfully."
                    })
            # fallback
            return JsonResponse({
                'is_success': False,
                'final_price': cls.format_price(plan.price if plan else 0),
                'price': format(plan.price if plan else 0, f".{Utility.get_val_by_name('decimal_number')}f"),
                'message': "This coupon code is invalid or has expired."
            })
        except PermissionDenied as err:
            return default_permission_denial(
                request,
                err=err,
                ref=REF,
                logger=logger,
                json={'error': 'Permission denied.'}
            )
        except Exception as err:
            return default_undefined_exception(
                request,
                err=err,
                ref=REF,
                logger=logger,
                json={'error': f"{CNAME}.{MNAME}: Error"},
                status=500
            )

    @classmethod
    def format_price(cls, price: float) -> str:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            admin_payment_setting = Utility.get_admin_payment_setting()
            currency = admin_payment_setting.get('currency', '$')
            decimal_number = Utility.get_val_by_name('decimal_number')
            return f"{currency}{format(price, f'.{decimal_number}f')}"
        except Exception as err:
            logger.error(f"{REF} Failed to format price: {err.__class__.__name__}: {err}")
            # TODO: consider raising or defaulting differently
            return f"${price}"

File: _inc/laravel/app/Http/Controllers/Bills/CouponController.php
<?php

namespace App\Http\Controllers\Bills;

use App\Http\Controllers\Abstracts\Controller;

use App\Config\Constants\{MiddlewaresConstants as MWC, PermissionsConstants as PMC, ViewsConstants as VW};
use App\Models\{Coupon, Plan, UserCoupon, Utility};
use Illuminate\Http\{Request, JsonResponse, RedirectResponse};
use Illuminate\Support\Facades\{Auth, Crypt, DB, Log, Route, Validator, View as ViewFacade};
use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};

final class CouponController extends Controller
{

    public function __construct()
    {
        $this->middleware([MWC::AUTH]);
    }

    public function index(Request $r): mixed
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $req    = $r;
        $viewPath = VW::CPN . '.index';
        return $this->measureProfile($action, function () use ($req, $action, $method, $class, $base, $viewPath) {
            Log::info("[{$base}::{$action}] start", ['user' => Auth::id(), 'method' => $method]);
            if (($auth = self::auth($req, PMC::MNG_CPN)) !== true) return $auth;
            try {
                $fetchStart = microtime(true);
                $coupons = Coupon::all();
                $this->logExecutionTime($fetchStart, $action, 'fetchCoupons');
                Log::info("[{$base}::{$action}] fetched", ['count' => $coupons->count()]);
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('coupons'));
                $this->logExecutionTime($renderStart, $action, 'renderIndex');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function create(Request $r): mixed
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $req    = $r;
        $viewPath = VW::CPN . '.create';
        return $this->measureProfile($action, function () use ($req, $action, $method, $class, $base, $viewPath) {
            Log::info("[{$base}::{$action}] start", ['user' => Auth::id(), 'method' => $method]);
            if (($auth = self::auth($req, 'create coupon')) !== true) return $auth;
            try {
                Log::info("[{$base}::{$action}] rendering form");
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath);
                $this->logExecutionTime($renderStart, $action, 'renderCreate');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function store(Request $r): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $req    = $r;
        return $this->measureProfile($action, function () use ($req, $action, $method, $class, $base) {
            Log::info("[{$base}::{$action}] start", ['user' => Auth::id(), 'method' => $method]);
            if (($auth = self::auth($req, 'create coupon')) !== true) return $auth;
            if ($resp = self::validateInput($req, ['name' => 'required', 'discount' => 'required|numeric', 'limit' => 'required|numeric'])) return $resp;
            if (!$req->manualCode && !$req->autoCode) {
                Log::warning("[{$base}::{$action}] missing code");
                return redirect()->back()->with('error', __('Coupon code is required'));
            }
            try {
                $createdId = null;
                $txnStart = microtime(true);
                DB::transaction(function () use ($req, $action, &$createdId) {
                    $createStart = microtime(true);
                    $coupon = new Coupon();
                    foreach (['name', 'discount', 'limit'] as $f) $coupon->$f = $req->input($f);
                    $coupon->code = $req->manualCode ? strtoupper($req->manualCode) : $req->autoCode;
                    $coupon->save();
                    $createdId = $coupon->id;
                    $this->logExecutionTime($createStart, $action, 'createCoupon');
                });
                $this->logExecutionTime($txnStart, $action, 'transaction');
                Log::info("[{$base}::{$action}] created", ['coupon_id' => $createdId]);
                return redirect()->route('coupons.index')->with('success', __('Coupon successfully created.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public function show(Coupon $coupon): mixed
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $viewPath = VW::CPN . '.view';
        return $this->measureProfile($action, function () use ($coupon, $action, $method, $class, $base, $viewPath) {
            Log::info("[{$base}::{$action}] start", ['coupon_id' => $coupon->id, 'method' => $method]);
            try {
                $fetchStart = microtime(true);
                $userCoupons = UserCoupon::where('coupon', $coupon->id)->with('userDetail')->get();
                $this->logExecutionTime($fetchStart, $action, 'fetchUserCoupons');
                Log::info("[{$base}::{$action}] fetched", ['count' => $userCoupons->count()]);
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('userCoupons'));
                $this->logExecutionTime($renderStart, $action, 'renderShow');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage(), 'coupon_id' => $coupon->id]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException(request(), $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'coupon_id' => $coupon->id]);
    }

    public function edit(Request $r, Coupon $coupon): mixed
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $req    = $r;
        $viewPath = VW::CPN . '.edit';
        return $this->measureProfile($action, function () use ($req, $coupon, $action, $method, $class, $base, $viewPath) {
            Log::info("[{$base}::{$action}] start", ['coupon_id' => $coupon->id, 'user' => Auth::id(), 'method' => $method]);
            if (($auth = self::auth($req, 'edit coupon')) !== true) return $auth;
            try {
                Log::info("[{$base}::{$action}] rendering", ['coupon_id' => $coupon->id]);
                if (!ViewFacade::exists($viewPath)) return back()->with('error', "HTTP 404: Page {$viewPath} not found!");
                $renderStart = microtime(true);
                $resp = view($viewPath, compact('coupon'));
                $this->logExecutionTime($renderStart, $action, 'renderEdit');
                return $resp;
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage(), 'coupon_id' => $coupon->id]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'coupon_id' => $coupon->id]);
    }

    public function update(Request $r, Coupon $coupon): RedirectResponse|JsonResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $req    = $r;
        return $this->measureProfile($action, function () use ($req, $coupon, $action, $method, $class, $base) {
            Log::info("[{$base}::{$action}] start", ['coupon_id' => $coupon->id, 'user' => Auth::id(), 'method' => $method]);
            if (($auth = self::auth($req, 'edit coupon')) !== true) return $auth;
            if ($resp = self::validateInput($req, ['name' => 'required', 'discount' => 'required|numeric', 'limit' => 'required|numeric', 'code' => 'required'])) return $resp;
            try {
                $updatedId = $coupon->id;
                $txnStart = microtime(true);
                DB::transaction(function () use ($req, $coupon, $action, &$updatedId) {
                    $updStart = microtime(true);
                    foreach (['name', 'discount', 'limit', 'code'] as $f) $coupon->$f = $req->input($f);
                    $coupon->save();
                    $this->logExecutionTime($updStart, $action, 'updateCoupon');
                    $updatedId = $coupon->id;
                });
                $this->logExecutionTime($txnStart, $action, 'transaction');
                Log::info("[{$base}::{$action}] updated", ['coupon_id' => $updatedId]);
                return redirect()->route('coupons.index')->with('success', __('Coupon successfully updated.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'coupon_id' => $coupon->id]);
    }

    public function destroy(Request $r, Coupon $coupon): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $req    = $r;
        return $this->measureProfile($action, function () use ($req, $coupon, $action, $method, $class, $base) {
            Log::info("[{$base}::{$action}] start", ['coupon_id' => $coupon->id, 'user' => Auth::id(), 'method' => $method]);
            if (($auth = self::auth($req, 'delete coupon')) !== true) return $auth;
            try {
                $deletedId = $coupon->id;
                $txnStart = microtime(true);
                DB::transaction(function () use ($coupon, $action, &$deletedId) {
                    $delStart = microtime(true);
                    $coupon->delete();
                    $this->logExecutionTime($delStart, $action, 'deleteCoupon');
                    $deletedId = $coupon->id ?? $deletedId;
                });
                $this->logExecutionTime($txnStart, $action, 'transaction');
                Log::info("[{$base}::{$action}] deleted", ['coupon_id' => $deletedId]);
                return redirect()->route('coupons.index')->with('success', __('Coupon successfully deleted.'));
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage(), 'coupon_id' => $coupon->id]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'coupon_id' => $coupon->id]);
    }

    public const AP_CPN = 'applyCoupon';
    public function applyCoupon(Request $r): JsonResponse|RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        $req    = $r;
        return $this->measureProfile($action, function () use ($req, $action, $method, $class, $base) {
            Log::info("[{$base}::{$action}] start", ['plan_id' => $req->plan_id, 'coupon' => $req->coupon, 'method' => $method]);
            try {
                $decStart = microtime(true);
                $planId = Crypt::decrypt($req->plan_id);
                $this->logExecutionTime($decStart, $action, 'decryptPlanId');
                $planFetch = microtime(true);
                $plan = Plan::find($planId);
                $this->logExecutionTime($planFetch, $action, 'fetchPlan');
                if (!$plan || !$req->coupon) {
                    Log::warning("[{$base}::{$action}] invalid input", ['plan_exists' => (bool)$plan, 'has_coupon' => (bool)$req->coupon]);
                    return response()->json(['is_success' => false]);
                }
                $origPrice = self::formatPrice($plan->price);
                $coupFetch = microtime(true);
                $coupon = Coupon::where('code', strtoupper($req->coupon))->where('is_active', '1')->first();
                $this->logExecutionTime($coupFetch, $action, 'fetchCoupon');
                if (!$coupon) {
                    Log::warning("[{$base}::{$action}] coupon not found");
                    return response()->json([
                        'is_success'  => false,
                        'final_price' => $origPrice,
                        'price'       => number_format($plan->price, Utility::getValByName('decimal_number')),
                        'message'     => __('This coupon code is invalid or has expired.')
                    ]);
                }
                if ($coupon->limit == $coupon->used_coupon()) {
                    Log::warning("[{$base}::{$action}] coupon exhausted", ['coupon_id' => $coupon->id]);
                    return response()->json([
                        'is_success'  => false,
                        'final_price' => $origPrice,
                        'price'       => number_format($plan->price, Utility::getValByName('decimal_number')),
                        'message'     => __('This coupon code has expired.')
                    ]);
                }
                $calcStart = microtime(true);
                $discount = ($plan->price / 100) * $coupon->discount;
                $final = $plan->price - $discount;
                $this->logExecutionTime($calcStart, $action, 'calculateDiscount');
                Log::info("[{$base}::{$action}] applied", ['coupon_id' => $coupon->id, 'discount' => $discount, 'final' => $final]);
                return response()->json([
                    'is_success'     => true,
                    'discount_price' => '-' . self::formatPrice($discount),
                    'final_price'    => self::formatPrice($final),
                    'price'          => number_format($final, Utility::getValByName('decimal_number')),
                    'message'        => __('Coupon code has applied successfully.')
                ]);
            } catch (\Throwable $e) {
                Log::error("[{$base}::{$action}] error", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return defaultUndefinedException($req, $e, $class . '::' . $action, '/', false, ['error' => __('Server error')]);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public const FMT_PRC = 'formatPrice';
    public const IDX = 'index';
    public const CRT = 'create';
    public const STR = 'store';
    public const SHW = 'show';
    public const EDT = 'edit';
    public const UPD = 'update';
    public const DEL = 'destroy';

    public function formatPrice(float|int $price): string
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        return $this->measureProfile($action, function () use ($price, $action, $method, $base) {
            Log::info("[{$base}::{$action}] start", ['price' => $price, 'method' => $method]);
            $setStart = microtime(true);
            $set = Utility::getAdminPaymentSetting();
            $this->logExecutionTime($setStart, $action, 'fetchSettings');
            $cur = $set['currency'] ?? '$';
            $fmtStart = microtime(true);
            $formatted = $cur . number_format($price, Utility::getValByName('decimal_number'));
            $this->logExecutionTime($fmtStart, $action, 'formatNumber');
            Log::info("[{$base}::{$action}] complete", ['formatted' => $formatted]);
            return $formatted;
        }, ['method' => $method, 'class' => $base]);
    }

    private static function auth(Request $r, string $perm): bool|RedirectResponse
    {
        Log::info(__CLASS__ . '::' . __FUNCTION__ . ' checking permission', [
            'user_id'    => $r->user()->id,
            'permission' => $perm,
        ]);
        if (!$r->user()->can($perm)) {
            Log::warning(__CLASS__ . '::' . __FUNCTION__ . ' denied', [
                'user_id'    => $r->user()->id,
                'permission' => $perm,
            ]);
            return defaultPermissionDenial(
                $r,
                new \Illuminate\Auth\Access\AuthorizationException(),
                __CLASS__ . '::' . debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1)[0]['function']
            );
        }
        Log::info(__CLASS__ . '::' . __FUNCTION__ . ' granted', [
            'user_id'    => $r->user()->id,
            'permission' => $perm,
        ]);
        return true;
    }

    private static function validateInput(Request $r, array $rules): ?RedirectResponse
    {
        Log::debug(__CLASS__ . '::' . __FUNCTION__ . ' validating input', [
            'rules' => $rules
        ]);
        $v = Validator::make($r->all(), $rules);
        if ($v->fails()) {
            $msg = $v->getMessageBag()->first();
            Log::warning(__CLASS__ . '::' . __FUNCTION__ . ' failed', ['error' => $msg]);
            return redirect()->back()->with('error', $msg);
        }
        Log::info(__CLASS__ . '::' . __FUNCTION__ . ' passed');
        return null;
    }
}
